File: src/main/java/org/example/kalkulationsprogramm/service/EmailOutboundPersistenceService.java
package org.example.kalkulationsprogramm.service;

import lombok.RequiredArgsConstructor;
import org.example.kalkulationsprogramm.domain.Email;
import org.example.kalkulationsprogramm.repository.EmailRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Propagation;
import org.springframework.transaction.annotation.Transactional;

/**
 * Persistiert eine selbst versandte (OUT) Email in einer separaten
 * Sub-Transaktion. Sinn: Aufrufer wie {@code AnfrageFunnelService} laufen
 * unter einer eigenen umschliessenden {@code @Transactional}, in der bereits
 * die fachlichen Daten (Anfrage, Notiz) angelegt wurden. Wuerde die
 * Email-Persistenz in derselben Transaktion scheitern (z.B.
 * {@code DataIntegrityViolationException} auf dem Unique-Index der
 * Message-ID, weil der IMAP-Sent-Scheduler die Mail Millisekunden vorher
 * importiert hat), markiert Hibernate die umfassende Transaktion als
 * rollback-only und der Funnel verliert den Lead — obwohl die SMTP-Mail
 * beim Kunden bereits angekommen ist.
 *
 * <p>Mit {@link Propagation#REQUIRES_NEW} laeuft jeder save in einer eigenen
 * Tx + Connection. {@code noRollbackFor=Exception.class} verhindert
 * Cross-Tx-Poisoning. Der Aufrufer schluckt den Fehler — der IMAP-Sent-Poll
 * legt die Mail spaeter ohnehin an (Sicherheitsnetz via Message-ID-Dedupe).
 */
@Service
@RequiredArgsConstructor
public class EmailOutboundPersistenceService {

    private final EmailRepository emailRepository;

    @Transactional(propagation = Propagation.REQUIRES_NEW, noRollbackFor = Exception.class)
    public void speichereOutEmail(Email email) {
        emailRepository.save(email);
    }

    /**
     * Auch der Dedupe-Lookup laeuft in einer eigenen Sub-Tx — ein Connection-Loss
     * hier darf die umfassende Funnel-Tx genauso wenig poisonieren wie der save.
     */
    @Transactional(propagation = Propagation.REQUIRES_NEW, readOnly = true, noRollbackFor = Exception.class)
    public boolean existsByMessageId(String messageId) {
        return emailRepository.existsByMessageId(messageId);
    }
}

File: src/main/kotlin/org/example/kalkulationsprogramm/service/EmailOutboundPersistenceService.kt
package org.example.kalkulationsprogramm.service

import org.example.kalkulationsprogramm.domain.Email
import org.example.kalkulationsprogramm.repository.EmailRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Propagation
import org.springframework.transaction.annotation.Transactional

@Service
class EmailOutboundPersistenceService(
    private val emailRepository: EmailRepository
) {
    @Transactional(propagation = Propagation.REQUIRES_NEW, noRollbackFor = [Exception::class])
    fun speichereOutEmail(email: Email) {
        emailRepository.save(email)
    }

    @Transactional(propagation = Propagation.REQUIRES_NEW, readOnly = true, noRollbackFor = [Exception::class])
    fun existsByMessageId(messageId: String): Boolean = emailRepository.existsByMessageId(messageId)
}
